diff --git a/src/torrent/utils/log.cc b/src/torrent/utils/log.cc
index 0c1c58936..8bbcc1b84 100644
--- a/src/torrent/utils/log.cc
+++ b/src/torrent/utils/log.cc
@@ -129,6 +129,18 @@ log_rebuild_cache() {
   }
 }
 
+void
+log_group::internal_print(const std::string& message) {
+  if (message.empty())
+    return;
+
+  auto lock = std::scoped_lock(log_mutex);
+
+  std::for_each(m_first, m_last, [this, &message](const auto& elem) {
+    return elem(message.c_str(), message.size(), std::distance(log_groups.begin(), this));
+  });
+}
+
 void
 log_group::internal_print(const HashString* hash, const char* subsystem, const void* dump_data, size_t dump_size, const char* fmt, ...) {
   va_list ap;
diff --git a/src/torrent/utils/log.h b/src/torrent/utils/log.h
index b9d04c0e1..91804147e 100644
--- a/src/torrent/utils/log.h
+++ b/src/torrent/utils/log.h
@@ -150,6 +150,7 @@ class LIBTORRENT_EXPORT log_group {
   // Internal:
   //
 
+  void                internal_print(const std::string& message);
   void                internal_print(const HashString* hash, const char* subsystem,
                                      const void* dump_data, size_t dump_size,
                                      const char* fmt, ...);
